File: cta-compliant-solution/source/resources/sdk/javascript/cta-client.js
/**
 * CTA-5007-B Native SDK - Local Token Generation
 * Generates real CBOR-encoded CWT tokens with COSE MAC0 structure
 * compatible with CloudFront's cf.cwt.validateToken()
 */

const crypto = require('crypto');

// --- Minimal CBOR encoder (sufficient for CWT/COSE structures) ---

function cborEncodeUint(value) {
    if (value < 24) return Buffer.from([value]);
    if (value < 0x100) return Buffer.from([0x18, value]);
    if (value < 0x10000) {
        const buf = Buffer.alloc(3);
        buf[0] = 0x19;
        buf.writeUInt16BE(value, 1);
        return buf;
    }
    if (value < 0x100000000) {
        const buf = Buffer.alloc(5);
        buf[0] = 0x1a;
        buf.writeUInt32BE(value, 1);
        return buf;
    }
    // 64-bit
    const buf = Buffer.alloc(9);
    buf[0] = 0x1b;
    buf.writeBigUInt64BE(BigInt(value), 1);
    return buf;
}

function cborEncodeNegInt(value) {
    // CBOR negative: -1 - n, so encode (abs(value) - 1) with major type 1
    const n = -value - 1;
    const encoded = cborEncodeUint(n);
    encoded[0] = (encoded[0] & 0x1f) | 0x20; // Set major type 1
    return encoded;
}

function cborEncodeInt(value) {
    if (value >= 0) return cborEncodeUint(value);
    return cborEncodeNegInt(value);
}

function cborEncodeBytes(buf) {
    const head = cborEncodeUint(buf.length);
    head[0] = (head[0] & 0x1f) | 0x40; // Major type 2
    return Buffer.concat([head, buf]);
}

function cborEncodeString(str) {
    const strBuf = Buffer.from(str, 'utf8');
    const head = cborEncodeUint(strBuf.length);
    head[0] = (head[0] & 0x1f) | 0x60; // Major type 3
    return Buffer.concat([head, strBuf]);
}

function cborEncodeArray(items) {
    const head = cborEncodeUint(items.length);
    head[0] = (head[0] & 0x1f) | 0x80; // Major type 4
    return Buffer.concat([head, ...items.map(cborEncode)]);
}

function cborEncodeMap(obj) {
    const keys = Object.keys(obj);
    const head = cborEncodeUint(keys.length);
    head[0] = (head[0] & 0x1f) | 0xa0; // Major type 5
    const parts = [head];
    for (const k of keys) {
        // Map keys: use integer if numeric, string otherwise
        const numKey = Number(k);
        if (!isNaN(numKey) && Number.isInteger(numKey)) {
            parts.push(cborEncodeInt(numKey));
        } else {
            parts.push(cborEncodeString(k));
        }
        parts.push(cborEncode(obj[k]));
    }
    return Buffer.concat(parts);
}

function cborEncodeTag(tag, value) {
    const head = cborEncodeUint(tag);
    head[0] = (head[0] & 0x1f) | 0xc0; // Major type 6
    const encoded = cborEncode(value);
    return Buffer.concat([head, encoded]);
}

/**
 * Wrap raw CBOR bytes in a CBOR tag without re-encoding the content.
 */
function cborWrapTag(tag, rawCbor) {
    const head = cborEncodeUint(tag);
    head[0] = (head[0] & 0x1f) | 0xc0; // Major type 6
    return Buffer.concat([head, rawCbor]);
}

function cborEncode(value) {
    if (value === null || value === undefined) return Buffer.from([0xf6]); // null
    if (typeof value === 'boolean') return Buffer.from([value ? 0xf5 : 0xf4]);
    if (typeof value === 'number') return cborEncodeInt(value);
    if (typeof value === 'string') return cborEncodeString(value);
    if (Buffer.isBuffer(value)) return cborEncodeBytes(value);
    if (Array.isArray(value)) return cborEncodeArray(value);
    if (typeof value === 'object') return cborEncodeMap(value);
    throw new Error(`Cannot CBOR encode: ${typeof value}`);
}

// --- COSE MAC0 / CWT token generation ---

/**
 * Build a COSE_MAC0 structure:
 *   Tag(17) [ protectedHeaders, unprotectedHeaders, payload, tag ]
 *
 * MAC structure for HMAC computation (per RFC 8152 §6.3):
 *   MAC_structure = ["MAC0", protectedHeaders, external_aad, payload]
 */
function buildCoseMac0(protectedHeaders, unprotectedHeaders, payload, key) {
    const protectedEncoded = cborEncode(protectedHeaders);
    const payloadEncoded = cborEncode(payload);

    // MAC_structure = ["MAC0", protectedEncoded, h'', payloadEncoded]
    const macStructure = cborEncodeArray([
        "MAC0",
        protectedEncoded,
        Buffer.alloc(0),  // external_aad
        payloadEncoded
    ]);

    // Compute HMAC-SHA256 tag
    const tag = crypto.createHmac('sha256', key).update(macStructure).digest();

    // COSE_MAC0 = [protectedEncoded, unprotectedHeaders, payloadEncoded, tag]
    const mac0Array = [protectedEncoded, unprotectedHeaders, payloadEncoded, tag];

    // Wrap in COSE_MAC0 tag (17) — use cborWrapTag since the array
    // contains pre-encoded CBOR buffers that must not be double-encoded
    const encodedArray = cborEncodeArray(mac0Array);
    return cborWrapTag(17, encodedArray);
}

/**
 * Generate a CWT token (optionally wrapped in CWT tag 61)
 */
function generateCWT(protectedHeaders, unprotectedHeaders, payload, key, options = {}) {
    const mac0 = buildCoseMac0(protectedHeaders, unprotectedHeaders, payload, key);

    if (options.cwtTag) {
        return cborWrapTag(61, mac0);
    }
    return mac0;
}

// --- CTA Client ---

class CTAClient {
    constructor(stackName, region = 'us-east-1') {
        this.stackName = stackName;
        this.region = region;
        this.keys = null;
    }

    async initSecretsManager(credentials = {}) {
        const { SecretsManagerClient, GetSecretValueCommand } = require("@aws-sdk/client-secrets-manager");
        this.smClient = new SecretsManagerClient({ region: this.region, ...credentials });
        this._GetSecretValueCommand = GetSecretValueCommand;
    }

    async getSigningKeys() {
        if (!this.smClient) throw new Error("Call initSecretsManager() first");

        const command = new this._GetSecretValueCommand({ SecretId: `${this.stackName}_CTAKey` });
        const response = await this.smClient.send(command);
        const secret = JSON.parse(response.SecretString);
        this.keys = { primary: { value: secret.signingKey, uuid: 'primary' } };
        return this.keys;
    }

    /**
     * Generate a real CBOR-encoded CWT token with COSE MAC0 structure.
     * The output is compatible with cf.cwt.validateToken().
     */
    generateCWTToken(policy, viewer = {}) {
        if (!this.keys) throw new Error("No signing keys. Call getSigningKeys() first");

        const now = Math.floor(Date.now() / 1000);
        const payload = {
            4: now + this.parseTTL(policy.ttl || '2h'), // exp
            5: now,  // nbf
            6: now   // iat
        };

        // URI restrictions: catu(312) → path(2) → prefix_match(1)
        if (policy.paths && policy.paths.length > 0) {
            payload[312] = { 2: { 1: policy.paths[0] } };
        }

        // Country restrictions: catgeoiso3166(316)
        if (policy.countries && policy.countries.length > 0) {
            payload[316] = policy.countries;
        }

        // Session ID: cti(7)
        if (policy.sessionId) {
            payload[7] = policy.sessionId;
        }

        const key = Buffer.from(this.keys.primary.value, 'base64');
        const protectedHeaders = { 1: 5 }; // alg: HMAC-SHA256
        const unprotectedHeaders = {};

        const tokenBuffer = generateCWT(protectedHeaders, unprotectedHeaders, payload, key, { cwtTag: true });
        const token = tokenBuffer.toString('base64url');

        return { token, claims: payload, expiresAt: payload[4] };
    }

    generateSignedUrl(mediaUrl, policy, viewer = {}) {
        const result = this.generateCWTToken(policy, viewer);

        if (policy.placement === 'query') {
            const separator = mediaUrl.includes('?') ? '&' : '?';
            return `${mediaUrl}${separator}CAT=${result.token}`;
        } else if (policy.placement === 'header') {
            return { url: mediaUrl, headers: { 'CTA-Common-Access-Token': result.token } };
        } else {
            const url = new URL(mediaUrl);
            return `${url.protocol}//${url.host}/${result.token}${url.pathname}${url.search}`;
        }
    }

    parseTTL(ttl) {
        if (typeof ttl === 'number') return ttl;
        const match = ttl.match(/^(\d+)([smhd])$/);
        if (!match) return 7200;
        const value = parseInt(match[1]);
        switch (match[2]) {
            case 's': return value;
            case 'm': return value * 60;
            case 'h': return value * 3600;
            case 'd': return value * 86400;
            default: return 7200;
        }
    }
}

if (typeof module !== 'undefined') module.exports = { CTAClient, generateCWT, cborEncode };
else window.CTAClient = CTAClient;

File: cta-compliant-solution/source/lambda-ruby/cta_client.rb
# CTA-5007-B Ruby SDK
# Generates COSE MAC0 / CWT tokens compatible with cf.cwt.validateToken()

require 'openssl'

module CTA
  # COSE / CWT constants (matching CloudFront docs)
  COSE_ALG = 1
  COSE_KID = 4
  HMAC_256 = 5

  module CWT
    ISS = 1; SUB = 2; AUD = 3; EXP = 4; NBF = 5; IAT = 6; CTI = 7
  end

  module CAT
    CATU = 401; CATNIP = 402; CATM = 403; CATR = 404
  end

  module CATU
    HOST = 1; PATH = 2; EXT = 3
  end

  module MATCH
    PREFIX = 1; SUFFIX = 2; EXACT = 3
  end

  # --- Minimal CBOR encoder ---

  def self.cbor_uint_head(major, value)
    m = major << 5
    if value < 24
      [m | value].pack('C')
    elsif value < 0x100
      [m | 24, value].pack('CC')
    elsif value < 0x10000
      [m | 25, value].pack('Cn')
    elsif value < 0x100000000
      [m | 26, value].pack('CN')
    else
      [m | 27, value].pack('CQ>')
    end
  end

  def self.cbor_encode(value)
    case value
    when Integer
      value >= 0 ? cbor_uint_head(0, value) : cbor_uint_head(1, -1 - value)
    when String
      if value.encoding == Encoding::BINARY || value.encoding == Encoding::ASCII_8BIT
        cbor_uint_head(2, value.bytesize) + value
      else
        bytes = value.encode('UTF-8')
        cbor_uint_head(3, bytes.bytesize) + bytes.b
      end
    when Array
      cbor_uint_head(4, value.length) + value.map { |v| cbor_encode(v) }.join
    when Hash
      cbor_uint_head(5, value.length) + value.map { |k, v| cbor_encode(k) + cbor_encode(v) }.join
    when NilClass
      "\xf6".b
    when TrueClass
      "\xf5".b
    when FalseClass
      "\xf4".b
    else
      raise "Cannot CBOR encode: #{value.class}"
    end
  end

  def self.cbor_tag(tag_num, content)
    if tag_num < 24
      [0xd8, tag_num].pack('CC') + content
    else
      cbor_uint_head(6, tag_num) + content
    end
  end

  # --- COSE MAC0 / CWT token generation ---

  def self.generate_token(claims, key, kid: 'key:default', cwt_tag: true)
    protected_bytes = cbor_encode({ COSE_ALG => HMAC_256 })
    unprotected_map = { COSE_KID => kid.encode('UTF-8').b }
    payload_bytes = cbor_encode(claims)

    mac_structure = cbor_encode(["MAC0", protected_bytes, "".b, payload_bytes])
    tag = OpenSSL::HMAC.digest('SHA256', key, mac_structure)

    arr = cbor_encode([protected_bytes, unprotected_map, payload_bytes, tag])
    cose_mac0 = cbor_tag(17, arr)

    cwt_tag ? cbor_tag(61, cose_mac0) : cose_mac0
  end

  def self.parse_ttl(ttl)
    return ttl if ttl.is_a?(Integer)
    m = ttl.to_s.match(/^(\d+)([smhd])$/)
    return 7200 unless m
    v = m[1].to_i
    { 's' => v, 'm' => v * 60, 'h' => v * 3600, 'd' => v * 86400 }[m[2]] || 7200
  end
end
